### PR TITLE
DODO 1632 not idempotent add_column when type differ

### PR DESCRIPTION
**Pull request by CharlesDelannoy**
- Raise error if try tro add a column that already exists with a different type
- Simplify message and add test

### CODE REVIEW

**Review by frederic-martin-doctolib — APPROVED** [review]

**Review by EQuincerot — COMMENTED** [review]
It would be nice to split the purpose of the PR with other changes :)
+ I think you should either get rid of the Helper.logger.say or only use it

File: test/StatementInsurer/change_column_null_test.rb [hunk 1]
```diff
@@ -126,7 +126,7 @@ def skip_if_unmet_requirements!
     end
 
     def met_requirements?
-      Gem::Requirement.new('>6.1').satisfied_by?(Gem::Version.new(::ActiveRecord::VERSION::STRING)) &&
+      Gem::Requirement.new('>6.1', '<7.1').satisfied_by?(Gem::Version.new(::ActiveRecord::VERSION::STRING)) &&
```
[the remark this answers is not in the document]
Comment (CharlesDelannoy) [reply test/StatementInsurer/change_column_null_test.rb:129]: We have issue with 7.1 in the tests. SO for now the dependency need to be lower than that
I'll create a ticket to have a look and improve compatibility and male it work with 7.1

File: lib/safe-pg-migrations/plugins/idempotent_statements.rb [hunk 2]
```diff
@@ -23,19 +20,24 @@ module IdempotentStatements
...
+      if column_exists?(table_name, column_name) && !column_exists?(table_name, column_name, type)
+        error_message = "/!\\ Column '#{column_name}' already exists in '#{table_name}' with a different type"
+        raise error_message
+      end
 
-      Helpers::Logger.say <<~MESSAGE.squish, sub_item: true
-        /!\\ Column '#{column_name}' already exists in '#{table_name}'. Skipping statement.
+      return super unless column_exists?(table_name, column_name, type)
+
+      log_message(<<~MESSAGE.squish
```
[the remark this answers is not in the document]
Comment (CharlesDelannoy) [reply lib/safe-pg-migrations/plugins/idempotent_statements.rb:30]: Rubocop was unhappy that there were too many lines in the module
I did a quick refactoring to satisfy him

File: .github/workflows/continuous-integration-workflow.yml [hunk 3]
```diff
@@ -31,7 +31,7 @@ jobs:
     # Service containers to run with `container-job`
     strategy:
       matrix:
-        postgres: [ 11.7, 12.9, "12.10" ]
+        postgres: [ 11.7, 12.14, 15.0 ]
```
[the remark this answers is not in the document]
Comment (CharlesDelannoy) [reply .github/workflows/continuous-integration-workflow.yml]: I put them both, If there is issue I'll only keep 15.2

File: lib/safe-pg-migrations/plugins/idempotent_statements.rb [hunk 4]
```diff
@@ -23,19 +20,24 @@ module IdempotentStatements
     end
 
     ruby2_keywords def add_column(table_name, column_name, type, *)
-      return super unless column_exists?(table_name, column_name)
+      if column_exists?(table_name, column_name) && !column_exists?(table_name, column_name, type)
```
Comment (fuyar) [inline lib/safe-pg-migrations/plugins/idempotent_statements.rb:23]: does this also check for properties like NOT NULL or DEFAULT ?

Comment (CharlesDelannoy) [reply lib/safe-pg-migrations/plugins/idempotent_statements.rb:23]: as it is not but I think that the method `column_exists` can do that
DO you want me to add those case also? 

If yes I think I will merge this PR and open another one if that's ok for you?

Comment (CharlesDelannoy) [reply lib/safe-pg-migrations/plugins/idempotent_statements.rb:23]: created a ticket to handle this later as discussed

**Pull request merged**